js/src/util/tokens/index.js

```diff
@@ -0,0 +1,293 @@
+          index: tokenIndex,
+
+          fetched: false
+        };
+      });
+    });
+}
+
+export function fetchTokensInfo (api, tokenReg, tokenIndexes) {
+  const requests = tokenIndexes.map((tokenIndex) => {
+    const tokenCalldata = tokenReg.getCallData(tokenReg.instance.token, {}, [tokenIndex]);
+
+    return { to: tokenReg.address, data: tokenCalldata };
+  });
+
+  const calls = requests.map((req) => api.eth.call(req));
+  const imagesPromise = fetchTokensImages(api, tokenReg, tokenIndexes);
+
+  let results;
+
+  return Promise.all(calls)
+    .then((_results) => {
+      results = _results;
+      return imagesPromise;
+    })
+    .then((images) => {
+      return results.map((rawTokenData, index) => {
+        const tokenIndex = tokenIndexes[index];
+        const tokenData = tokenReg.instance.token
+          .decodeOutput(rawTokenData)
+          .map((t) => t.value);
+
+        const [ address, tag, format, name ] = tokenData;
+        const image = images[index];
+
+        const token = {
+          address,
+          id: sha3(address + tokenIndex).slice(0, 10),
+          index: tokenIndex,
+
+          format: format.toString(),
+          image,
+          name,
+          tag,
+
+          fetched: true
+        };
+
+        return token;
+      });
+    });
+}
+
+export function fetchTokensImages (api, tokenReg, tokenIndexes) {
+  const requests = tokenIndexes.map((tokenIndex) => {
+    const metaCalldata = tokenReg.getCallData(tokenReg.instance.meta, {}, [tokenIndex, 'IMG']);
+
+    return { to: tokenReg.address, data: metaCalldata };
+  });
+
+  const calls = requests.map((req) => api.eth.call(req));
+
+  return Promise.all(calls)
+    .then((results) => {
+      return results.map((rawImage) => {
+        const image = tokenReg.instance.meta.decodeOutput(rawImage)[0].value;
+
+        return hashToImageUrl(image);
+      });
+    });
+}
+
+/**
+ * `updates` should be in the shape:
+ *   {
+ *     [ who ]: [ tokenId ]  // Array of tokens to updates
+ *   }
+ *
+ * Returns a Promise resolved with the balances in the shape:
+ *   {
+ *     [ who ]: { [ tokenId ]: BigNumber } // The balances of `who`
+ *   }
+ */
+export function fetchAccountsBalances (api, tokens, updates) {
+  const accountAddresses = Object.keys(updates);
+
+  // Updates for the ETH balances
+  const ethUpdates = accountAddresses
+    .map((accountAddress) => {
+      return updates[accountAddress].filter((tokenId) => tokenId === ETH_TOKEN.id);
+    })
+    .reduce((nextUpdates, tokenIds, accountIndex) => {
+      if (tokenIds.length > 0) {
+        const accountAddress = accountAddresses[accountIndex];
+
+        nextUpdates[accountAddress] = tokenIds;
+      }
+
+      return nextUpdates;
+    }, {});
+
+  // Updates for Tokens balances
+  const tokenUpdates = Object.keys(updates)
+    .map((accountAddress) => {
+      return updates[accountAddress].filter((tokenId) => tokenId !== ETH_TOKEN.id);
```

Comment:
This could be simplified as well:
```js
const tokenUpdates = Object.keys(updates).reduce((nextUpdates, address) => {
   const tokens = updates[address].filter(tokenId => tokenId !== ETH_TOKEN.id);
   if (tokens.length) {
      nextUpdates[address] = tokens;
   }
   return nextUpdates;
}, {});
```